snuba/datasets/storages/selectors/selector.py

```diff
@@ -0,0 +1,77 @@
+from abc import ABC, abstractmethod
+from typing import List, Sequence, Type, cast
+
+from snuba.datasets.storage import (
+    ReadableStorage,
+    StorageAndMappers,
+    StorageAndMappersNotFound,
+)
+from snuba.datasets.storages.factory import get_storage
+from snuba.datasets.storages.storage_key import StorageKey
+from snuba.query.logical import Query
+from snuba.query.query_settings import QuerySettings
+from snuba.utils.registered_class import RegisteredClass
+
+
+class QueryStorageSelectorError(Exception):
+    pass
+
+
+class QueryStorageSelector(ABC, metaclass=RegisteredClass):
+    """
+    The component provided by a dataset and used at the beginning of the
+    execution of a query to pick the storage query should be executed onto.
+    """
+
+    @classmethod
+    def config_key(cls) -> str:
+        return cls.__name__
+
+    @classmethod
+    def get_from_name(cls, name: str) -> Type["QueryStorageSelector"]:
+        return cast(Type["QueryStorageSelector"], cls.class_from_name(name))
+
+    @abstractmethod
+    def select_storage(
+        self,
+        query: Query,
+        query_settings: QuerySettings,
+        storage_and_mappers_list: List[StorageAndMappers],
+    ) -> StorageAndMappers:
+        raise NotImplementedError
+
+    def get_storage_mapping_pair(
+        self,
+        storage: ReadableStorage,
+        storage_and_mappers_list: Sequence[StorageAndMappers],
+    ) -> StorageAndMappers:
+        for sm_tuple in storage_and_mappers_list:
+            if storage == sm_tuple.storage:
```

Comment:
This check feels a bit brittle. The problem is that we don't have a proper equality check for storages, so what you are actually checking here is that the instance of the storages is the same (Python compares the memory address of the variables and sees if they are the same).

In theory that should be fine, since we only instantiate each storage once, but it would be better to compare the storage key I think. Storages are aware of their own storage key. `storage.storage_key == ...`

Reply:
Makes sense, updated to compare using storage key instead.